projects/ping_delays.kp/knowledge.md

```diff
@@ -0,0 +1,253 @@
+
+Specifically investigating what typical values of "recording delay" and "submission delay" might be.
+
+
+```python
+import ujson as json
+import matplotlib.pyplot as plt
+import pandas as pd
+import numpy as np
+import plotly.plotly as py
+import IPython
+
+from datetime import datetime, timedelta
+from email.utils import parsedate_tz, mktime_tz, formatdate
+
+from plotly.graph_objs import *
+from moztelemetry import get_pings_properties, get_one_ping_per_client
+from moztelemetry.dataset import Dataset
+
+%matplotlib inline
+IPython.core.pylabtools.figsize(16, 7)
+```
+    Unable to parse whitelist (/home/hadoop/anaconda2/lib/python2.7/site-packages/moztelemetry/histogram-whitelists.json). Assuming all histograms are acceptable.
+
+
+Start with the bread-and-butter of Telemetry reporting: the "main" ping
+
+Looking at Jan 10, 2017 because it's a recent Tuesday that isn't too close to any holidays.
+
+
+```python
+pings = Dataset.from_source("telemetry") \
+    .where(docType='main') \
+    .where(submissionDate="20170110") \
+    .records(sc, sample=0.01)
+```
+To look at delays, we need to look at times. There are a lot of times, and they are recorded relative to different clocks.
+
+**`creationDate`** - The time the Telemetry code in Firefox created the ping, according to the client's clock, expressed as an ISO string. **`meta/creationTimestamp`** is the same time, but expressed in nanoseconds since the epoch.
+
+**`meta/Date`** - The time the Telemetry code in Firefox sent the ping to the server, according to the client's clock, expressed as a Date string conforming to [RFC 7231](https://tools.ietf.org/html/rfc7231#section-7.1.1.1).
+
+**`meta/Timestamp`** - The time the ping was received by the server, according to the server's
+clock, expressed in nanoseconds since the epoch.
+
+**`payload/info/subsessionLength`** - The length of time over which the ping was collecting data, according to the client's clock, expressed as a number of seconds. Working backwards from **`creationDate`** this gives us a subsession start time which allows us to determine Reporting Delay.
+
+
+```python
+subset = get_pings_properties(pings, ["application/channel",
+                                      "creationDate",
+                                      "meta/creationTimestamp",
+                                      "meta/Date",
+                                      "meta/Timestamp",
+                                      "payload/info/subsessionLength"])
+```
+
+```python
+p = subset.take(1)[0]
+```
+
+```python
+p
+```
+
+
+
+
+    {'application/channel': u'release',
+     'creationDate': u'2017-01-10T02:01:31.551Z',
+     'meta/Date': u'Tue, 10 Jan 2017 02:01:31 GMT',
+     'meta/Timestamp': 1484013691737682688L,
+     'meta/creationTimestamp': 1.484013691551e+18,
+     'payload/info/subsessionLength': 67}
+
+
+
+Quick normalization: ditch any ping that doesn't have a subsessionLength, creationTimestamp, or Timestamp:
```

Comment:
Very few. The last time I did this, across some 3-4million pings I was dropping a couple dozen. So... less than 0.01%?